### PR TITLE
[mlir] Change `tensor.extract/insert` to take static/dynamic indices.

### PR DESCRIPTION
**Pull request by cathyzhyi**
This changes the ODS of `tensor.extract/insert` op. Some new builder methods are added and the verifiers/canonicalizers are updated. One of the canonicalization pattern of `shape.shape_of` is also updated.

### CODE REVIEW

File: mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td [hunk 1]
```diff
@@ -332,12 +332,37 @@ def Tensor_ExtractOp : Tensor_Op<"extract", [
...
+  }];
+
+  let builders = [
+    // Build an ExtractOp with mixed static and dynamic indexes.
+    OpBuilder<(ins "Value":$tensor, "ArrayRef<OpFoldResult>":$indexes,
+      CArg<"ArrayRef<NamedAttribute>", "{}">:$attrs)>,
+    // Build an ExtractOp with mixed static, dynamic indexes and inferred result type.
+    OpBuilder<(ins "Type":$resultType, "Value":$tensor, "ArrayRef<OpFoldResult>":$indexes,
+      CArg<"ArrayRef<NamedAttribute>", "{}">:$attrs)>,
+    // Build an ExtractOp with dynamic indexes.
```
Comment (matthias-springer) [inline mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:359]: There are some spelling inconsistences: `indexes`, `indices`

File: mlir/lib/Dialect/Shape/IR/Shape.cpp [hunk 2]
```diff
@@ -1736,6 +1736,32 @@ struct ShapeOfFromReshape : public OpRewritePattern<shape::ShapeOfOp> {
...
+
+  LogicalResult matchAndRewrite(tensor::ExtractOp op,
+                                PatternRewriter &rewriter) const override {
+    auto tensorShapeOfOp = op.getTensor().getDefiningOp<shape::ShapeOfOp>();
+    if (!tensorShapeOfOp)
+      return rewriter.notifyMatchFailure(op, "producer is not shape.shape_of");
+
+    int64_t staticIndice = op.getStaticIndices()[0];
+    Type indexType = rewriter.getIndexType();
+    Value indice =
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Shape/IR/Shape.cpp:1751]: typo: `indices`

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp [hunk 3]
```diff
@@ -1137,12 +1189,18 @@ OpFoldResult ExtractOp::fold(FoldAdaptor adaptor) {
...
-  for (Attribute indice : adaptor.getIndices()) {
-    if (!indice || !llvm::isa<IntegerAttr>(indice))
-      return {};
-    indices.push_back(llvm::cast<IntegerAttr>(indice).getInt());
+  auto dynamicIndicesIt = adaptor.getIndices().begin();
+  for (int64_t i : getStaticIndices()) {
+    if (i != ShapedType::kDynamic) {
+      indices.push_back(i);
+    } else {
+      Attribute indice = *dynamicIndicesIt;
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:1197]: typo: `indices`

**Review by matthias-springer — COMMENTED** [review]
There are a few more places that have to be updated. Basically everything that calls `ExtractOp::getIndices` or `InsertOp::getIndices`. E.g., `Tensor/Transforms/BufferizableOpInterfaceImpl.cpp`.

File: mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td [hunk 4]
```diff
@@ -808,16 +833,35 @@ def Tensor_InsertOp : Tensor_Op<"insert", [
...
   let results = (outs AnyRankedTensor:$result);
   let assemblyFormat = [{
-    $scalar `into` $dest `[` $indices `]` attr-dict `:` type($dest)
+    $scalar `into`
+    $dest `` custom<DynamicIndexList>($indices, $static_indices)
+    attr-dict `:` type($dest)
   }];
 
   let extraClassDeclaration = [{
     MutableOperandRange getDpsInitsMutable() { return getDestMutable(); }
```
[the remark this answers is not in the document]
Comment (cathyzhyi) [reply mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:847]: Thanks for the suggestion! Do you think a new interface like MixedIndicesInterface is needed for this?

[the remark this answers is not in the document]
Comment (matthias-springer) [reply mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:847]: That could be useful. (But can also be done without.) I tried something like that in the past (https://reviews.llvm.org/D156899), but I didn't land it for some reason... Don't really remember why. There was also an RFC (https://discourse.llvm.org/t/rfc-more-opfoldresult-and-mixed-indices-in-ops-that-deal-with-shaped-values/72510). I would read through that discussion before adding a new interface.

[the remark this answers is not in the document]
Comment (jpienaar) [reply mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:847]: +1 I think there is something more general to be done here, I think this is a good starting point and can see other pain points and think about generalizing.

File: mlir/lib/Dialect/Shape/IR/Shape.cpp (hunk 2 above)
Comment (jpienaar) [reply mlir/lib/Dialect/Shape/IR/Shape.cpp:1751]: Rather index ?

File: mlir/lib/Dialect/Shape/IR/Shape.cpp [hunk 5]
```diff
@@ -1736,6 +1736,32 @@ struct ShapeOfFromReshape : public OpRewritePattern<shape::ShapeOfOp> {
...
+        staticIndice != ShapedType::kDynamic
```
Comment (jpienaar) [inline mlir/lib/Dialect/Shape/IR/Shape.cpp:1752]: Prefer using isDynamic(staticIndex)

File: mlir/lib/Dialect/Shape/IR/Shape.cpp [hunk 6]
```diff
@@ -1736,6 +1736,32 @@ struct ShapeOfFromReshape : public OpRewritePattern<shape::ShapeOfOp> {
...
+    : public OpRewritePattern<tensor::ExtractOp> {
+  using OpRewritePattern<tensor::ExtractOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExtractOp op,
+                                PatternRewriter &rewriter) const override {
+    auto tensorShapeOfOp = op.getTensor().getDefiningOp<shape::ShapeOfOp>();
+    if (!tensorShapeOfOp)
+      return rewriter.notifyMatchFailure(op, "producer is not shape.shape_of");
+
+    int64_t staticIndice = op.getStaticIndices()[0];
```
Comment (jpienaar) [inline mlir/lib/Dialect/Shape/IR/Shape.cpp:1749]: Its sort of weird that static index could be dynamic ... I seem to recall poking about this on a previous review, why not just store only static in one and only dynamic in the other and then using the type to differentiate - that would result in more operations for indexing. Not something to address here as this is keeping the form.

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp [hunk 7]
```diff
@@ -39,6 +39,19 @@ using llvm::divideCeilSigned;
 using llvm::divideFloorSigned;
 using llvm::mod;
 
+static LogicalResult
+checkTensorRankMatchIndices(Value tensor, ValueRange dynamicIndices,
+                            ArrayRef<int64_t> staticIndices) {
+  auto tensorType = llvm::cast<RankedTensorType>(tensor.getType());
+  int64_t dynamicDimCount = llvm::count_if(staticIndices, [](int64_t element) {
+    return element == ShapedType::kDynamic;
```
Comment (jpienaar) [inline mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:47]: Same

File: mlir/lib/Dialect/Shape/IR/Shape.cpp (hunk 2 above)
Comment (jpienaar) [reply mlir/lib/Dialect/Shape/IR/Shape.cpp:1751]: (funnily enough it seems indice is index in Spanish)

**Review by joker-eph — CHANGES_REQUESTED** [review]
What is the motivation for this?

In general I see this as an anti-pattern to have operation accept both an attribute and an SSA value.
I thought the "experiment" with the few ops that adopted this form in the past was deemed like an accepted "bad idea" in the end?

**Comment by MaheshRavishankar** [comment]
> What is the motivation for this?
> 
> In general I see this as an anti-pattern to have operation accept both an attribute and an SSA value.
> I thought the "experiment" with the few ops that adopted this form in the past was deemed like an accepted "bad idea" in the end?

I think there is a long history of discussion here, but for one I prefer the mixed static/dynamic list. It is pretty readable and constant values are folded into the operation itself instead have to lookup use-def chains.  So I would very much prefer having a general utility of the form being added in this PR. So +1 from me!

**Comment by joker-eph** [comment]
To be clear: it has nothing to do with readability but instead more with the uniformity of the APIs (or lack thereof) and the complexity imposed on every user.

**Comment by MaheshRavishankar** [comment]
> To be clear: it has nothing to do with textual readability but instead more with the uniformity of the APIs (or lack thereof) and the complexity imposed on every user.

Could you explain more. It's opt-in based on the op. So what is the complexity you are referring to?

**Comment by joker-eph** [comment]
>  It's opt-in based on the op.

I don't quite get what you mean by this?

>  So what is the complexity you are referring to?

The C++ API provides accessors for either Value or Attribute which describe the same thing and are both optional (one of them must be present, in an exclusionary way).
MLIR core infra is just not setup to support this use case very well (we could have improved it to do this, but it isn't a pattern that has been widely accepted and so it hasn't happened). For example you can use getOperands() to get a `ValueRange`, and ODS can generate accessors for `ValueRange` for named operands, but this "idiom" is not correctly abstracted anywhere in the same way and the complexity is pushed on every API users.

**Comment by MaheshRavishankar** [comment]
> >  It's opt-in based on the op.
> 
> I don't quite get what you mean by this?

I mean it only matters to the op that wants to use this paradigm. So in that sense it is opt-in. 

> 
> >  So what is the complexity you are referring to?
> 
> The C++ API provides accessors for either Value or Attribute which describe the same thing and are both optional (one of them must be present, in an exclusionary way).

Not sure that is true. As I understand it the attribute is always required. The Values are needed only if the attribute has a sentinel value saying the dimension is dynamic.

> MLIR core infra is just not setup to support this use case very well (we could have improved it to do this, but it isn't a pattern that has been widely accepted and so it hasn't happened). For example you can use getOperands() to get a `ValueRange`, and ODS can generate accessors for `ValueRange` for named operands, but this "idiom" is not correctly abstracted anywhere in the same way and the complexity is pushed on every API users.

That is true. Most ops that use this use custom methods to get mixed static/dynamic values.  This PR IMO is starting to build support for this paradigm and make it's usage more uniform

**Comment by joker-eph** [comment]
> Most ops that use this use custom methods to get mixed static/dynamic values. This PR IMO is starting to build support for this paradigm and make it's usage more uniform

So far you motivated this by “it is nicer to read”: I would strongly object to doing anything this scale for this reason. I am sure there are others though but I’d want to see this well motivated (why this op and not just all operations?), and carefully considered, because this isn’t trivial.

in the past this pattern started with other motivations (@nicolasvasilache or @ftynse ), but the tradeoffs didn’t pan out as far as I can remember.

**Comment by matthias-springer** [comment]
Another point that wasn't mentioned yet: Static indices can be verified. Dynamic indices should not be verified as per verifier guidelines.

**Comment by MaheshRavishankar** [comment]
> > Most ops that use this use custom methods to get mixed static/dynamic values. This PR IMO is starting to build support for this paradigm and make it's usage more uniform
> 
> So far you motivated this by “it is nicer to read”: I would strongly object to doing anything this scale for this reason. I am sure there are others though but I’d want to see this well motivated (why this op and not just all operations?), and carefully considered, because this isn’t trivial.
> 
> in the past this pattern started with other motivations (@nicolasvasilache or @ftynse ), but the tradeoffs didn’t pan out as far as I can remember.

Agreed that "nicer to read" isnt a good enough reason (but is a good side-effect). For me its more about things that are constant, like constant values in list of offsets, sizes and strides of slices, (or indices in this PR) are "embedded" into the operation itself. I dont need to look at producer of operands to see if it is a constant (and as Matthias mentioned, verify these constants values).  So thats why +1 from me on this. 

basically
```
%0 = tensor.extract %b[10, 20]
```

is more self-contained

```
%c10 = arith.constant 10 : index
%c20 = arith.constant 20 : index
%0 = tensor.extract %b[%c10, %c20]
```

and also allows you to verify that `[10, 20]` is in-bounds of `%b` without having to look at producers of operands.

**Comment by joker-eph** [comment]
> For me its more about things that are constant,

Either they are constant and always attributes, or they are "maybe" constant and that's an SSA value that you can match. It's the same with every single operation though as far as I can tell. I don't see what's special here.
(hence my comment about if we were intended for this, we'd have built this in the core infra and generalized it as a first class thing).

**Comment by MaheshRavishankar** [comment]
> > For me its more about things that are constant,
> 
> Either they are constant and always attributes, or they are "maybe" constant and that's an SSA value that you can match. It's the same with every single operation though as far as I can tell. I don't see what's special here.
> (hence my comment about if we were intended for this, we'd have built this in the core infra and generalized it as a first class thing).

Sure... What would first class support for this look like. Could we build it in now?

**Comment by joker-eph** [comment]
> Sure... What would first class support for this look like. Could we build it in now?

I don't know what it would look like (something like: replace `Value` with `OpFoldResult` for all API on the `Operation` class to begin with? Changes to ODS?). 
Right now MLIR is just designed as "constant are defined with separate operations", this not a simple change and it would need a very strong motivation.
(attributes are not means to fold SSA value, they are meant for information **required** by lowering)

**Comment by jpienaar** [comment]
I don't see this as proposing a change to MLIR core. This is a change to the Tensor dialect, making the insert and extract ops more uniform within the dialect itself.

I currently have inputs where I have O(40k) scalar constant operations whose only reason for existing is indexing. Its inefficient, its more difficult to test and cumbersome to write (using a little lazy cache per isolated from above context to avoid ending at O(400k) scalar constant operations that get CSE'd later). Its not special to indexing ops that one could capture constant values, but it is very wasteful for them (I forget, its 10x higher storage than just parameter).

We could also just introduce new ops: tensor.insert\_static and tensor.extract\_static that is all static. That is an option. It introduces different set of conditional "expansion" (e.g., where insert is expected today, one would need to consider the static insert too).

Now wrt API, indeed this change doesn't change those. An option, is to add method like `std::optional<int64_t> getConstantDim(i)` which would then either look at constant property or match on operand to try and extract constant value, and `std::optional<Value> getDynamicDim(i)` which would just return operand else nullopt. Still results in conditionals, but for static that's not new (match is already a conditional) and unseen by user, while for dynamic I actually don't know how its being used except for codegen later and there it is most likely post checking if constant already :) This also doesn't do any automatic promotion to static args, so its only if created such.

**Comment by joker-eph** [comment]
> I don't see this as proposing a change to MLIR core

Nobody said otherwise I believe.

> We could also just introduce new ops: tensor.insert_static and tensor.extract_static that is all static

I don’t have an objection of principles on this if there is a use-case for static indexing.

**Comment by matthias-springer** [comment]
What do you think about `OpBuilder::createOrFold`? It currently returns `Value`, but I think it should return `OpFoldResult`. I tried to change that at some point (but it was too much work). As a follow-up, I wanted to update more tensors ops to store "mixed" values. (Similar to what this PR is doing.) Then the API would work nicely with `createOrFold`. (In a sense that the op builder can take the result of a `createOrFold` and does not have to build an op for a constant value.) But also, that never happened...
